stuffer/s2n_stuffer.c

```diff
@@ -146,6 +146,7 @@ int s2n_stuffer_wipe_n(struct s2n_stuffer *stuffer, const uint32_t size)
     stuffer->write_cursor -= size;
     memset_check(stuffer->blob.data + stuffer->write_cursor, S2N_WIPE_PATTERN, size);
     stuffer->read_cursor = MIN(stuffer->read_cursor, stuffer->write_cursor);
+    stuffer->high_water_mark = MIN(stuffer->high_water_mark, stuffer->write_cursor);
```

Comment:
@camshaft any suggestions here?

Reply:
I think we can only do that if `write_cursor + size >= high_water_mark`. Consider the following:

```
[1, 2, 3, 0]
    ^ write cursor
       ^ high water mark

s2n_stuffer_wipe_n(stuffer, 1);

[0, 2, 3, 0]
 ^ write cursor
       ^ high water mark <- needs to stay this high
```

Reply:
Thinking about it more: this function always wipes from right to left, since it unwinds the `write_cursor` by `size` before wiping `size` bytes. So the only time it would make sense to set the `high_water_mark = write_cursor` would be if, before subtracting `size`,  `write_cursor == high_water_mark` (i.e. on this line it would be `write_cursor == high_water_mark - size`).

Reply:
If we resize a stuffer to a smaller size, we are currently updating all the cursors, but the `high_water_mark`. So there is a case where the `high_water_mark` points to the portion we wiped when shrinking the stuffer. This is not a problem on the `wipe_n`, but on the `resize` function itself (`wipe_n` is not updating `blob.size`). I changed `resize` to properly update the `high_water_mark`. Do you still think it's necessary to cover the case you mentioned on the `wipe_n` function? @camshaft

Reply:
Proposed change. What do you think @camshaft 
```
int s2n_stuffer_wipe_n(struct s2n_stuffer *stuffer, const uint32_t size)
{
    if (size >= stuffer->write_cursor) {
        return s2n_stuffer_wipe(stuffer);
    }

    /* We know that size is now less than write_cursor */
    stuffer->write_cursor -= size;
    memset_check(stuffer->blob.data + stuffer->write_cursor, S2N_WIPE_PATTERN,(stuffer->high_water_mark - stuffer->write_cursor));
    stuffer->high_water_mark = stuffer->write_cursor;
    stuffer->read_cursor = MIN(stuffer->read_cursor, stuffer->write_cursor);

    return S2N_SUCCESS;
}
```